src/libsync/creds/httpcredentials.cpp

```diff
@@ -239,44 +239,54 @@ bool HttpCredentials::stillValid(QNetworkReply *reply)
                || !reply->property(authenticationFailedC).toBool()));
 }
 
-bool HttpCredentials::refreshAccessToken()
+bool HttpCredentials::refreshAccessTokenInternal(int tokenRefreshRetriesCount)
 {
     if (_refreshToken.isEmpty())
         return false;
     if (_isRenewingOAuthToken) {
         return true;
     }
-    if (++_tokenRefreshRetriesCount >= TokenRefreshMaxRetries) {
-        qCWarning(lcHttpCredentials) << "Too many failed refreshs" << _tokenRefreshRetriesCount << "-> log out";
-        forgetSensitiveData();
-        Q_EMIT _account->invalidCredentials();
-        _tokenRefreshRetriesCount = 0;
-        return false;
-    }
     _isRenewingOAuthToken = true;
 
     // don't touch _ready or the account state will start a new authentication
     // _ready = false;
 
     OAuth *oauth = new OAuth(_account, this);
-    connect(oauth, &OAuth::refreshError, this, [oauth, this](QNetworkReply::NetworkError error, const QString &errorString) {
+    connect(oauth, &OAuth::refreshError, this, [oauth, tokenRefreshRetriesCount, this](QNetworkReply::NetworkError error, const QString &) {
         oauth->deleteLater();
-        _isRenewingOAuthToken = false;
+        int nextTry = tokenRefreshRetriesCount + 1;
+        std::chrono::seconds timeout = {};
         switch (error) {
         case QNetworkReply::ContentNotFoundError:
             // 404: bigip f5?
-            QTimer::singleShot(0, this, &HttpCredentials::refreshAccessToken);
+            timeout = 0s;
             break;
+        case QNetworkReply::HostNotFoundError:
+            Q_FALLTHROUGH();
+        case QNetworkReply::TimeoutError:
+            Q_FALLTHROUGH();
+        case QNetworkReply::TemporaryNetworkFailureError:
+            nextTry = 0;
+            Q_FALLTHROUGH();
         default:
-            QTimer::singleShot(30s, this, &HttpCredentials::refreshAccessToken);
+            timeout = 30s;
+        }
+        if (nextTry >= TokenRefreshMaxRetries) {
+            qCWarning(lcHttpCredentials) << "Too many failed refreshs" << nextTry << "-> log out";
+            forgetSensitiveData();
+            Q_EMIT _account->invalidCredentials();
+            return;
         }
+        QTimer::singleShot(timeout, this, [nextTry, this] {
+            _isRenewingOAuthToken = false;
```

Comment:
I think that part was essential too as else different parts of the client could retrigger it independent of the timer.

Reply:
Sounds reasonable.